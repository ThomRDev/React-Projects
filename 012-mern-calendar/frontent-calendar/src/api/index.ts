export {default as calendarApi} from './calendarApi'
// export * as calendarApi from './calendarApi'