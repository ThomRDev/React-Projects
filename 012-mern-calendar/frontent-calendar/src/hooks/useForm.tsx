import { useState, ChangeEvent, useEffect, useMemo } from 'react';
export const useForm = <T extends Object> (initialState: T,formValidations : any = {}) => {
  const [formState,setForm] = useState(initialState)
  const [formValidation, setFormValidation] = useState<any>({})

  // cada ves que cualquier campo del form cambie, verificarÃ¡ las validaciones
  useEffect(()=>{
    createValidators()
  },[ formState ])

  // si cambia los parametros
  useEffect(()=>{ 
    setForm(initialState)
  },[initialState])

  const onInputChange = (event:ChangeEvent<HTMLInputElement>) => {
    const { name,value } = event.target
    setForm({
      ...formState,
      [name] : value
    })
  }

  const onReset = () => setForm(initialState)

  const createValidators = ()=>{
    const formCheckValues:any = {}
    for (const formField in formValidations) {
      if (Object.prototype.hasOwnProperty.call(formValidations, formField)) {
        const formFieldCapitalize = formField.charAt(0).toUpperCase() + formField.slice(1);
        const [validatorFunction,errorMessage = "This field is required"] = formValidations[formField]
        formCheckValues[`is${formFieldCapitalize}Valid`] = validatorFunction(formState[formField as keyof typeof formState]) ? null : errorMessage
      }
    }
    setFormValidation(formCheckValues)
  }

  const isFormValid = useMemo(()=>{
    for (const formField in formValidation) {
      if (Object.prototype.hasOwnProperty.call(formValidations, formField)) {
        if(formValidation[formField] != null) return false
      }
    }
    return true
  },[formValidation])

  return {
    ...formState,
    formState,
    onInputChange,
    onReset,
    formValidation,
    isFormValid
  }
}