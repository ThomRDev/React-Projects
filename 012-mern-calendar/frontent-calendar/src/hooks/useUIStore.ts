import { useSelector } from "react-redux"
import { onCloseDateModal, onOpenDateModal, RootState, useAppDispatch } from "../store"

export const useUIStore = () => {
  const dispatch = useAppDispatch()

  const { isDateModalOpen } = useSelector((state:RootState)=>state.ui)

  const openDateModal = () => {
    dispatch(onOpenDateModal())
  }

  const closeDateModal = () => {
    dispatch(onCloseDateModal())
  }

  return {
    isDateModalOpen,

    // metodos
    openDateModal,
    closeDateModal
  }

}
