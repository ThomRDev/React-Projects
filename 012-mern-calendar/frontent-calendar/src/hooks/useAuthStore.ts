import { useSelector } from "react-redux";
import { calendarApi } from "../api";
import {
  clearErrorMessage,
  onChecking,
  onLogin,
  onLogoutCalendar,
  onLogoutWithError,
  onLogoutWithoutError,
  RootState,
  useAppDispatch,
} from "../store";
import { AxiosError } from "axios";

interface login {
  email: string;
  password: string;
}
interface register extends login {
  name: string;
}

export const useAuthStore = () => {
  const dispatch = useAppDispatch();
  const { status, user, errorMessage } = useSelector(
    (root: RootState) => root.auth
  );

  const startLogin = async (formLogin: login) => {
    dispatch(onChecking());
    try {
      const { data } = await calendarApi.post("/auth/login", formLogin);
      localStorage.setItem("token", data.token);
      dispatch(onLogin({ name: data.user.name, uid: data.user.uid }));
    } catch (error) {
      const err = error as AxiosError;
      const errorResponse = err.response?.data as any;
      if(!errorResponse){
        console.log('entre');
        dispatch(onLogoutWithError(err.message));
      }else{
        if (!!errorResponse?.msg) {
          dispatch(onLogoutWithError(errorResponse?.msg));
        } else if (!!errorResponse.errors) {
          const arrErrors = Object.values(errorResponse.errors).map(
            (e: any) => e.msg
          );
          dispatch(onLogoutWithError(arrErrors.join(",")));
        } else {
          dispatch(onLogoutWithError("Credenciales incorrectas"));
        }
      }
      setTimeout(() => {
        dispatch(clearErrorMessage());
      }, 10);
    }
  };

  const startRegister = async ({ email, password, name }: register) => {
    dispatch(onChecking());
    try {
      const { data } = await calendarApi.post("/auth/register", {
        email,
        password,
        name,
      });
      localStorage.setItem("token", data.token);
      dispatch(onLogin({ name: data.user.name, uid: data.user.id }));
    } catch (error) {
      const err = error as AxiosError;
      const errorResponse = err.response?.data as any;
      if (!!errorResponse?.msg) {
        dispatch(onLogoutWithError(errorResponse?.msg));
      } else if (!!errorResponse.errors) {
        const arrErrors = Object.values(errorResponse.errors).map(
          (e: any) => e.msg
        );
        dispatch(onLogoutWithError(arrErrors.join(",")));
      } else {
        dispatch(onLogoutWithError("Credenciales incorrectas"));
      }
      setTimeout(() => {
        dispatch(clearErrorMessage());
      }, 10);
    }
  };

  const startLogout = () => {
    localStorage.clear();
    dispatch(onLogoutWithoutError());
    dispatch(onLogoutCalendar());
  };

  // cuando entro por primera ves, si hay un token que vuelva a generar un nuevo token
  const checkAuthToken = async () => {
    const token = localStorage.getItem("token");
    if (!token) return dispatch(onLogoutWithoutError());
    try {
      const { data } = await calendarApi.get("/auth/refresh-token");
      localStorage.setItem("token", data.token);
      dispatch(onLogin({ name: data.user.name, uid: data.user.uid }));
    } catch (err) {
      localStorage.clear();
      dispatch(onLogoutWithoutError());
    }
  };

  return {
    status,
    user,
    errorMessage,
    startLogin,
    startRegister,
    startLogout,
    checkAuthToken,
  };
};
