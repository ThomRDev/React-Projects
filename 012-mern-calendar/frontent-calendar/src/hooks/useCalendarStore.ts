import { AxiosError } from 'axios';
import { useSelector } from "react-redux";
import Swal from "sweetalert2";
import { calendarApi } from '../api';
import { convertEventsToDateEvents } from '../helpers/convertEventsToDateEvents';
import { errorHandler } from '../helpers/errorHandler';
import {
  ACTIVE_EVENT_TYPE,
  calendarSlice,
  Event,
  onAddNewEvent,
  onClearActiveEvent,
  onDeleteEvent,
  onLoadEvents,
  onSetActiveEvent,
  onUpdateEvent,
  RootState,
  TYPES_OF_ACTIVE_EVENTS,
  useAppDispatch,
} from "../store";

export const useCalendarStore = () => {
  const dispatch = useAppDispatch();
  const { events, activeEvent } = useSelector(
    (state: RootState) => state.calendar
  );

  const { user } = useSelector((state: RootState) => state.auth)

  const setActiveEvent = (calendarEvent: ACTIVE_EVENT_TYPE) => {
    dispatch(onSetActiveEvent(calendarEvent));
  };

  const clearActiveNote = () => {
    dispatch(onClearActiveEvent());
  };

  const startSavingEvent = async (calendarEvent: Event) => {
    const payload = {
      title:calendarEvent.title,
      start:calendarEvent.start,
      end:calendarEvent.end,
      notes:calendarEvent.notes
    }
    try {
      if (calendarEvent.id) {
        await calendarApi.put(`/events/${calendarEvent.id}`,payload)
        dispatch(onUpdateEvent({...calendarEvent,user} as Event));
      } else {
        const { data } = await calendarApi.post(`/events`,payload)
        dispatch(
          onAddNewEvent({ ...calendarEvent, id: data.event.id,user } as Event)
        );
      }
    } catch (error) {
      errorHandler(error as AxiosError,'Error al guardar','No se puedo guardar el evento')
    }
  };

  const startDeletingEvent = async () => {
    try {
      await calendarApi.delete(`/events/${ (activeEvent?.event as Event).id }` );
      dispatch( onDeleteEvent() );
    } catch (error) {
      console.log(error);
      errorHandler(error as AxiosError,'Error al eliminar','No se puedo eliminar el evento')
    }
  }
  
  const startLoadingEvents = async () => {
    try{
      const {data} = await calendarApi.get('/events')
      console.log(data.events)
      // se tiene que formatear el date ya que no es compatible
      // ya que el servidor devuelve string y no del tipo Date
      // el react-big-calendar necesita un objeto de tipo Date
      dispatch(onLoadEvents(convertEventsToDateEvents(data.events)))
    }catch(error){
      errorHandler(error as AxiosError,'Error al cargar los eventos','No se puedo cargar los eventos')
    }
  }
  

  return {
    events,
    activeEvent,
    setActiveEvent,
    clearActiveNote,
    startSavingEvent,
    startDeletingEvent,
    startLoadingEvents,
    
    // si se selecciono
    hasEventSelected: !!activeEvent && activeEvent.type === TYPES_OF_ACTIVE_EVENTS.EXISTING_EVENT,
  };
};
