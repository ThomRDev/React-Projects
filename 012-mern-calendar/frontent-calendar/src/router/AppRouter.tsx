import { useEffect } from "react"
import { Navigate, Route, Routes } from "react-router-dom"
import { LoginPage } from "../auth"
import { CalendarPage } from "../calendar"
import { useAuthStore } from "../hooks"
import { AUTH_STATE, AUTH_STATUS } from "../store"
import { Loader } from "../ui/components/Loader"



export const AppRouter = () => {
  // const authStatus:AUTH_STATE = AUTH_STATUS.NOT_AUTHENTICATED

  const { status:authStatus,checkAuthToken } = useAuthStore()

  useEffect(()=>{
    checkAuthToken()
  },[])

  if(authStatus === AUTH_STATUS.CHECKING) {
    // return <p>Loading ...</p>
    return <Loader />
  }

  return (
    <Routes>
      {
        authStatus === AUTH_STATUS.NOT_AUTHENTICATED ? 
        <>
          {/* <Route path="/auth/*" element={<LoginPage />} />  */}
          <Route path="/auth/login" element={<LoginPage />} /> 
          <Route path="/*" element={<Navigate to="/auth/login" />} /> 
        </>:   
        <>
          <Route path="/" element={<CalendarPage />} />
          <Route path="/*" element={ <Navigate to="/" /> } />
        </>
      }
      
      
    </Routes>
  )
}
