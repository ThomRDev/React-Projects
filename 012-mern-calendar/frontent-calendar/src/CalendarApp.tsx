import { Provider } from "react-redux"
import { BrowserRouter, HashRouter } from "react-router-dom"
import { AppRouter } from "./router"
import { store } from "./store"

export const CalendarApp = () => {
  return (
    <Provider store={store}>
      {/* una solucion para el path en un servidor */}
      {/* es usar el * en node con express y redirijir a index.html */}
      {/* otra es usar el # */}
      {/* 
        BrowserRouter /auth/login
        HashRouter  /#/auth/login
      */}
      {/* <HashRouter> */}
        <BrowserRouter>
          <AppRouter />
        </BrowserRouter>
      {/* </HashRouter> */}
    </Provider>
  )
}
// https://www.reddit.com/r/reactjs/comments/q6lit4/thoughts_about_redux_reactquery_together/