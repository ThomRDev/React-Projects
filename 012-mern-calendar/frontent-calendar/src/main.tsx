import React from 'react'
import ReactDOM from 'react-dom/client'
import { CalendarApp } from './CalendarApp'
import './index.css'

ReactDOM.createRoot(document.getElementById('root') as HTMLElement).render(
  // calendar me obliga a quitar el modo estricto (UNSAFE_componentWillReceiveProps )
  // <React.StrictMode>
    <CalendarApp />
  // </React.StrictMode>
)
