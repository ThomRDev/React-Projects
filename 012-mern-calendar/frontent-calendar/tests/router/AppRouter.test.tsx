import { render, screen } from "@testing-library/react"
import { MemoryRouter } from "react-router-dom";
import { useAuthStore } from "../../src/hooks/useAuthStore";
import { AppRouter } from "../../src/router";
import { AUTH_STATUS } from "../../src/store/auth/authSlice";

// tengo que colocar esto si me cuasa varios problemas
jest.mock('../../src/calendar', () => ({
  CalendarPage: () => <h1>CalendarPage</h1>
}))

jest.mock('../../src/hooks/useAuthStore');
const mockUseAuthStore = useAuthStore as jest.MockedFunction<typeof useAuthStore>;


describe('Test in <AppRouter />', () => {
  const mockCheckAuthToken = jest.fn();
  beforeEach(() => jest.clearAllMocks() );
  test('should show the loading screen and call checkAuthToken', () => {

    mockUseAuthStore.mockReturnValue({
      status: AUTH_STATUS.CHECKING,
      checkAuthToken:mockCheckAuthToken,
    } as any)

    render(
      <AppRouter />
    )
    expect( screen.getByText('Cargando') ).toBeTruthy() 
    expect( mockCheckAuthToken ).toHaveBeenCalled();
  })
  test('should show the login page if not authenticated', () => {

    mockUseAuthStore.mockReturnValue({
      status: AUTH_STATUS.NOT_AUTHENTICATED,
      checkAuthToken:mockCheckAuthToken,
    } as any)

    const { container } = render(
      <MemoryRouter initialEntries={['/calendar']}>
        <AppRouter />
      </MemoryRouter>
    )
    expect( screen.getByText('Ingreso') ).toBeTruthy();
    expect( container ).toMatchSnapshot(); 
  })
  test('should show the calendar page if authenticated', () => {

    mockUseAuthStore.mockReturnValue({
      status: AUTH_STATUS.AUTHENTICATED,
      checkAuthToken:mockCheckAuthToken,
    } as any)

    const { container } = render(
      <MemoryRouter initialEntries={['/calendar']}>
        <AppRouter />
      </MemoryRouter>
    )
    expect( screen.getByText('CalendarPage') ).toBeTruthy();
  })
})

/**
el problema con testing library es que no puedo probar librerias de terceros
un componente de terceros, testing library si lo renderiza a div div o lo que sea
enzime si me lo construia, colocaba el componente como tal <MyComponent />

El paquete de terceros no se hacen pruebas
 */