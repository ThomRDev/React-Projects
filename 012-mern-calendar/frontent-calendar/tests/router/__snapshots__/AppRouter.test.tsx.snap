// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test in <AppRouter /> should show the login page if not authenticated 1`] = `
<div>
  <div
    class="container login-container animate__animated animate__fadeIn"
    style="margin: auto;"
  >
    <div
      class="row"
    >
      <div
        class="col-md-6 login-form-1"
      >
        <h3>
          Ingreso
        </h3>
        <form>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="email"
              placeholder="Correo"
              type="text"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="password"
              placeholder="Contraseña"
              type="password"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="btnSubmit"
              type="submit"
              value="Login"
            />
          </div>
        </form>
      </div>
      <div
        class="col-md-6 login-form-2"
      >
        <h3>
          Registro
        </h3>
        <form>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="name"
              placeholder="Nombre"
              type="text"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="email"
              placeholder="Correo"
              type="email"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="password"
              placeholder="Contraseña"
              type="password"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="confirmPassword"
              placeholder="Repita la contraseña"
              type="password"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="btnSubmit"
              type="submit"
              value="Crear cuenta"
            />
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;
